lib/mayaUsd/fileio/primUpdaterManager.cpp

```diff
@@ -1724,4 +1747,270 @@ bool PrimUpdaterManager::readPullInformation(const MDagPath& dagPath, Ufe::Path&
     return false;
 }
 
+/* static */
+std::list<PrimUpdaterManager::VariantSetDescriptor>
+PrimUpdaterManager::variantSetDescriptors(const Ufe::Path& p)
+{
+    std::list<VariantSetDescriptor> vsd;
+    auto                            path = p;
+    while (path.nbSegments() > 1) {
+        auto ancestor = Ufe::Hierarchy::createItem(path);
+        auto usdAncestor = std::static_pointer_cast<MayaUsd::ufe::UsdSceneItem>(ancestor);
+        auto variantSets = usdAncestor->prim().GetVariantSets();
+        std::list<VariantSelection> vs;
+        for (const auto& vsn : variantSets.GetNames()) {
+            vs.emplace_back(vsn, variantSets.GetVariantSelection(vsn));
+        }
+        vsd.emplace_back(path, vs);
+        path = path.pop();
+    }
+    return vsd;
+}
+
+void PrimUpdaterManager::beginManagePulledPrims()
+{
+    TF_VERIFY(_pulledPrims.root()->empty());
```

Comment:
Isn't beginManagedPulledPrims() called unconditionally and this the observer might already exist if we are editing a second USD prim? I would have expected this function to start with a check that _orphanedNodesManager is null or not and do nothing if it already exists. Would be also more foolproof if for some reason it gets called twice in any case.